keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,345 @@
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a new component of Kubelet, it guarantees NUMA aware memory usage for the Pod which has Guarantedd QoS, here memory includes hugepages.
+
+To guarantee NUMA aware memory usage, when Guaranteed QoS Pod admission is requested, Memory Manager caculates NUMA node affinity to figure out appropriate NUMA node where enough memory and hugepages exist for containers in the Pod. Then, when Pod deployed, it enfoces containers in the Pod to consume memory and hugepages from designated NUMA node which Topology Manager selected.
+
+NOTE:
+- Memory Manager isolates Memory and Hugepages to single NUMA node to prevent remote(Inter-NUMA) memory access. For this, Memory Manager enforce cgroups cpuset subsystem using CRI API. For more details, see [CPUSET Enforcement](#cpuset-enforcement).
+
+- Memory Manager guarantees that the container(in Guanteed QoS Pod) consumes requested amount of memory(which is isolated to a NUMA node) through linux OOM system. For more details, see [OOM Score Enforcement](#oom-score-enforcement)
+
+More details of this component are listed below.
+
+#### The Concept of Safety Zone
+
+Memory Manager has the concept of safety zone to manage memory and hugepages at the NUMA node level. Basically, Kubernetes support reserving memory at the node level through Node Allocatable feature. It provides options such as `system-reserved` and `kube-reserved` to reserve CPUs, Memory, Storage for system services and kubernetes system components. Fortunately, it is possible to reserve CPUs with consideration of NUMA because CPUs are reserved using their logical core number. On the other hand, it is impossible to reserve memory with consideration of NUMA, the only thing we can do is reserving memory at the node level.
+
+The concept of safety zone is proposed to reserve certain amount of memory and hugepages equally on each NUMA node. The value of safety zone can be confiugred by configuration parameter of Memory Manager. For more details see the [Feature Gate and Kubelet Flags](#feature-gate-and-kubelet-flags) section.
+- For memory, total amount of sefety zone memory(value of safety zone memory * number of NUMA nodes) cannot be greater than total amount of reserved memory(kube-reserved, system-reserved and eviction-threshold) by Node Allocatable feature.
```

Comment:
So is the safety zone considered to be *part of* the reserved memory, or is it *in addition to* the reserved memory?  (Same for hugepages.)

Maybe consider making the answer to this explicit in the document?

Reply:
`Maybe consider making the answer to this explicit in the document?`
=> sure, I also got some review from coleague, I wil consider it together.

The fomular to dertermine allocatable size of memory is:
`total memory of NUMA node` - `Safety Zone`

So it will not be part of reserved memory, reserved memory means amount of reserved for memory for container in Guaranteed QoS Class Pod from allocatable zone.


```
           +---------------+------------+--------------+
           |               |       Allocatable         |
NUMA node  +  Safety Zone  +------------+--------------+
           |               |  Reserved  |     free     |
           +---------------+------------+--------------+
```

Reply:
I was sure, by reserved here, we are meaning reserved memory for kube daemons and OS tasks, how does it connect to `Guaranteed QoS Class Pod`?

Reply:
Oh, I think I made a mistake. :(
I just confused between `reserved` of Allocatable Zone and `reserved memory` of node allocatable feature.
Guys, please forget my above answer.


` total amount of sefety zone memory(value of safety zone memory * number of NUMA nodes) cannot be greater than total amount of reserved memory(kube-reserved, system-reserved and eviction-threshold) by Node Allocatable feature.`
The reserved memory on the above sentence means the reserved memory by node allocatable feature at the node level.

Answer for @cbf123 is: 
Yes, the safety zone will be part of reserved memory(of node allocatable feature).

For hugepages, node allocatable feature doesn't have any mechanism to reserve hugepages not.
so Memory Manager won't have a safety zone before Node allocatable feature support reserving hugepages.
I write about it right below, see 146 line.
```
- For hugepages, at the moment(v1.17), Node Allocateble features does not support reserving hugepages, so that Memory Manager will not have safety zone for hugepages. Once Node Allocatable feature will support hugepages reservation([#83541](https://github.com/kubernetes/kubernetes/pull/83541)), safety zone for hugapages will works in the same manner of memory.
```